OpenRA.Mods.Common/Pathfinder/PathCacheStorage.cs

```diff
@@ -9,12 +9,77 @@
  */
 #endregion
 
+using System;
 using System.Collections.Generic;
 using System.Linq;
 
 namespace OpenRA.Mods.Common.Pathfinder
 {
-	public class PathCacheStorage : ICacheStorage<List<CPos>>
+	public enum PathCacheQueryType : byte
+	{
+		UnitPath,
+		UnitPathToRange
+	}
+
+	public readonly struct PathCacheKey
+	{
+		readonly PathCacheQueryType queryType;
+		readonly uint actorID;
+		readonly int source;
+		readonly int target;
+		readonly int targetY;
+		readonly int hash;
+
+		public PathCacheKey(PathCacheQueryType queryType, uint actorID, CPos source, CPos target)
+		{
+			this.queryType = queryType;
+			this.actorID = actorID;
+			this.source = source.Bits;
+			this.target = target.Bits;
+			targetY = -1;
+			hash = HashCode.Combine<PathCacheQueryType, uint, int, int>(queryType,
```

Comment:
The types can be inferred by the compiler, so this can simplify to
```csharp
hash = HashCode.Combine(queryType, actorID, this.source, this.target);
```